src/test/Main.hs

```diff
@@ -1,162 +1,428 @@
-    Witch.from @B.ByteString @LB.ByteString x == LB.fromStrict x
-
-  Hspec.prop "From LazyByteString ByteString" $ \ x ->
-    Witch.from @LB.ByteString @B.ByteString x == LB.toStrict x
-
-  Hspec.prop "From String Text" $ \ x ->
-    Witch.from @String @T.Text x == T.pack x
-
-  Hspec.prop "From Text String" $ \ x ->
-    Witch.from @T.Text @String x == T.unpack x
-
-  Hspec.prop "From Text LazyText" $ \ x ->
-    Witch.from @T.Text @LT.Text x == LT.fromStrict x
-
-  Hspec.prop "From LazyText Text" $ \ x ->
-    Witch.from @LT.Text @T.Text x == LT.toStrict x
-
-  Hspec.prop "From [a] (Seq a)" $ \ x ->
-    Witch.from @[Int] @(Seq.Seq Int) x == Seq.fromList x
-
-  Hspec.prop "From (Seq a) [a]" $ \ x ->
-    Witch.from @(Seq.Seq Int) @[Int] x == Foldable.toList x
-
-  Hspec.prop "From [a] (Set a)" $ \ x ->
-    Witch.from @[Int] @(Set.Set Int) x == Set.fromList x
-
-  Hspec.prop "From (Set a) [a]" $ \ x ->
-    Witch.from @(Set.Set Int) @[Int] x == Set.toAscList x
-
-  Hspec.prop "From [(k, v)] (Map k v)" $ \ x ->
-    Witch.from @[(Char, Int)] @(Map.Map Char Int) x == Map.fromList x
-
-  Hspec.prop "From (Map k v) [(k, v)]" $ \ x ->
-    Witch.from @(Map.Map Char Int) @[(Char, Int)] x == Map.toAscList x
-
-instance QuickCheck.Arbitrary Natural.Natural where
-  arbitrary = QuickCheck.arbitrarySizedNatural
-  shrink = QuickCheck.shrinkIntegral
-
-instance QuickCheck.Arbitrary a => QuickCheck.Arbitrary (NonEmpty.NonEmpty a) where
-  arbitrary = QuickCheck.applyArbitrary2 (NonEmpty.:|)
-  shrink = QuickCheck.genericShrink
-
-instance QuickCheck.Arbitrary LB.ByteString where
-  arbitrary = fmap LB.pack QuickCheck.arbitrary
-  shrink = QuickCheck.shrinkMap LB.pack LB.unpack
-
-instance QuickCheck.Arbitrary B.ByteString where
-  arbitrary = fmap B.pack QuickCheck.arbitrary
-  shrink = QuickCheck.shrinkMap B.pack B.unpack
-
-instance QuickCheck.Arbitrary T.Text where
-  arbitrary = fmap T.pack QuickCheck.arbitrary
-  shrink = QuickCheck.shrinkMap T.pack T.unpack
-
-instance QuickCheck.Arbitrary LT.Text where
-  arbitrary = fmap LT.pack QuickCheck.arbitrary
-  shrink = QuickCheck.shrinkMap LT.pack LT.unpack
+main = runTestTTAndExit $ "Witch" ~:
+  [ as @Int8 1 ~?= 1
+  , cast (1 :: Int8) ~?= (1 :: Int16)
+  , from @Int8 1 ~?= (1 :: Int)
+  , into @Int16 (1 :: Int8) ~?= 1
+  , over @String (<> "!") (Name "Simon") ~?= Name "Simon!"
+  , via @Int16 (1 :: Int8) ~?= (1 :: Int32)
+  , tryFrom @Int16 1 ~?= Right (1 :: Int8)
+  , tryInto @Int8 (1 :: Int16) ~?= Right 1
+  , unsafeCast (1 :: Int16) ~?= (1 :: Int8)
+  , unsafeFrom @Int16 1 ~?= (1 :: Int8)
+  , unsafeInto @Int8 (1 :: Int16) ~?= 1
+
+# if MIN_VERSION_GLASGOW_HASKELL(8, 10, 1, 0)
+  , ($$( liftedCast (1 :: Int16) ) :: Int8) ~?= 1
+  , ($$( liftedFrom @Int16 1 ) :: Int8) ~?= 1
+  , ($$( liftedInto @Int8 (1 :: Int16) ) :: Int8) ~?= 1
+# else
+  , $( liftedCast @Int16 @Int8 1 ) ~?= (1 :: Int8)
+  , $( liftedFrom @Int16 @Int8 1 ) ~?= (1 :: Int8)
+  , $( liftedInto @Int8 @Int16 1 ) ~?= (1 :: Int8)
```

Comment:
The ergonomics of these untyped Template Haskell splices are pretty bad. They need to be fully specified because type inference doesn't go "into" or "from" the splice. For example, this is ambiguous:

``` hs
-- ambiguous
$( liftedFrom @Int16 1 ) :: Int8
-- needs to be
$( liftedFrom @Int16 @Int8 1 ) :: Int8
```

Even though the result must be an `Int8`, that information doesn't make its way to the inside of the splice. Similar, this is ambiguous:

``` hs
-- ambiguous
$( liftedCast @Int16 @Int8 1 ) ~?= 1
-- needs to be
$( liftedCast @Int16 @Int8 1 ) ~?= (1 :: Int8)
```

Even though the result has to be an `Int8`, that information doesn't make its way out of the splice.

Reply:
Ah, I didn't realize that GHC 8.8 supported typed Template Haskell splices like `$$(...)` even though it doesn't have `liftTyped`. Should be fixed now. 

This also makes documentation easier: Everything uses `$$(...)` regardless of GHC version.